cache/cache.go

```diff
@@ -0,0 +1,928 @@
+		DuplicateFiles:          false, // storage doesn't permit this
+		ReadMimeType:            wrappedFs.Features().ReadMimeType,
+		WriteMimeType:           wrappedFs.Features().WriteMimeType,
+		CanHaveEmptyDirectories: wrappedFs.Features().CanHaveEmptyDirectories,
+		BucketBased:             wrappedFs.Features().BucketBased,
+		Purge:                   f.Purge,
+		Copy:                    f.Copy,
+		Move:                    f.Move,
+		DirMove:                 f.DirMove,
+		DirChangeNotify:         nil,
+		DirCacheFlush:           f.DirCacheFlush,
+		PutUnchecked:            f.PutUnchecked,
+		CleanUp:                 f.CleanUp,
+		UnWrap:                  f.UnWrap,
+	}).Fill(f).Mask(wrappedFs)
+
+	return f, wrapErr
+}
+
+// Name of the remote (as passed into NewFs)
+func (f *Fs) Name() string {
+	return f.name
+}
+
+// Root of the remote (as passed into NewFs)
+func (f *Fs) Root() string {
+	return f.root
+}
+
+// Features returns the optional features of this Fs
+func (f *Fs) Features() *fs.Features {
+	return f.features
+}
+
+// String returns a description of the FS
+func (f *Fs) String() string {
+	return fmt.Sprintf("%s:%s", f.name, f.root)
+}
+
+// Cache is the persistent type cache
+func (f *Fs) Cache() Storage {
+	return f.cache
+}
+
+// ChunkSize returns the configured chunk size
+func (f *Fs) ChunkSize() int64 {
+	return f.chunkSize
+}
+
+// OriginalSettingWorkers will return the original value of this config
+func (f *Fs) OriginalSettingWorkers() int {
+	return f.originalTotalWorkers
+}
+
+// OriginalSettingChunkNoMemory will return the original value of this config
+func (f *Fs) OriginalSettingChunkNoMemory() bool {
+	return f.originalChunkMemory
+}
+
+// InWarmUp says if cache warm up is active
+func (f *Fs) InWarmUp() bool {
+	return f.warmUp
+}
+
+// EnableWarmUp will enable the warm up state of this cache along with the relevant settings
+func (f *Fs) EnableWarmUp() {
+	f.totalWorkers = 1
+	f.chunkMemory = false
+	f.warmUp = true
+}
+
+// DisableWarmUp will disable the warm up state of this cache along with the relevant settings
+func (f *Fs) DisableWarmUp() {
+	f.totalWorkers = f.OriginalSettingWorkers()
+	f.chunkMemory = !f.OriginalSettingChunkNoMemory()
+	f.warmUp = false
+}
+
+// NewObject finds the Object at remote.
+func (f *Fs) NewObject(remote string) (fs.Object, error) {
+	co := NewObject(f, remote)
+	err := f.cache.GetObject(co)
+	if err == nil {
+		return co, nil
+	}
+	obj, err := f.Fs.NewObject(remote)
+	if err != nil {
+		return nil, err
+	}
+	co = ObjectFromOriginal(f, obj)
+	co.Persist()
+	return co, nil
+}
+
+// List the objects and directories in dir into entries
+func (f *Fs) List(dir string) (entries fs.DirEntries, err error) {
+	// clean cache
+	go f.CleanUpCache(false)
+
+	cd := NewDirectory(f, dir)
+	entries, err = f.Cache().GetDirEntries(cd)
+	if err != nil {
+		fs.Debugf(dir, "no dir entries in cache: %v", err)
+	} else if len(entries) == 0 {
+		// TODO: read empty dirs from source?
```

Comment:
No. I've left a TODO there to remind me in the future to approach this case differently.

Directories can be cached in 2 ways:
1. as part of a listing of its parent directory - here it will be cached with no info on what it contains
2. as part of a request to be listed - here it will be cached along with its listing

You can see that when one directory gets cached from 1 first and then we request a list of that directory it will be empty in the cache. That's why for now, we have to always check empty directories at the source.

I guess an easy way to overcome this would be to add a flag to a cached dir to track this but the current impact is low. I just thought of this by the way, this edge is there from day 1 and I left it as it is.

Reply:
I see.  And what happens if the directory is empty?

Reply:
It exits the if and requests a `List` from the wrapped FS. It's basically always a cache miss on empty folders.

Reply:
Not ideal, but OK!

Reply:
I agree. It's not ideal. It should be something addressed in the next revision